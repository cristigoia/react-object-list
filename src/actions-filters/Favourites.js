import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ClassNames from 'classnames'
import FavouritesItem from './FavouritesItem'

class Favourites extends Component {
  static propTypes = {
    /** list of favourites available to select from */
    favourites: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
      filters: PropTypes.shape({
        activeFilters: PropTypes.object, // {filterKey: filterValue} eg. {status: '&status=ACTIVE'}
        optionalFields: PropTypes.object, // {fieldName1: true, fieldName2: true} - value always seem to be true
        activeSort: PropTypes.string, // eg. '-id'
      }),
    })),
    /** callback function when deleting a favourite */
    handleDeleteFavourite: PropTypes.func,
    /** the name of the currently selected favourite name */
    selectedFavouriteName: PropTypes.string,
    /** callback function when adding a new favourite */
    handleAddFavourite: PropTypes.func,
    /** callback function to set the selected favourite to the current favourite */
    loadFavourite: PropTypes.func,
    /** Icon to render next to favourites */
    FavouritesIcon: PropTypes.element,
    /** Icon to render to remove a favourite */
    RemoveFavouriteIcon: PropTypes.element,
    /** Icon to click to expand favourites dropdown */
    OpenIcon: PropTypes.element,
    /** Icon to click to close favourites dropdown */
    CloseIcon: PropTypes.element,
  }

  static defaultProps = {
    favourites: [],
    selectedFavouriteName: null,
    FavouritesIcon: <i className="fa fa-heart-o" />,
    OpenIcon: <i className="fa fa-caret-down" />,
    CloseIcon: <i className="fa fa-caret-up" />,
  }

  state = {
    favouritesOpen: false,
    newFavouriteName: '',
    selectedFavouriteName: this.props.selectedFavouriteName,
  }

  componentWillMount() {
    document.body.addEventListener('click', this.handleDropdown)
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.handleDropdown)
  }

  componentWillReceiveProps(nextProps) {
    const { selectedFavouriteName } = nextProps
    if (selectedFavouriteName !== this.state.selectedFavouriteName) {
      this.setState({selectedFavouriteName})
    }
  }

  /**
   * Toggles open status of favourites dropdown
   */
  handleDropdown = (event) => {
    this.setState(prevState => {
      const newState = {
        favouritesOpen: false,
      }
      if (event.target === this.favouritesButton) {
        newState.favouritesOpen = !prevState.favouritesOpen
      } else if (event.target) {
        let el = event.target.parentElement
        while (el) {
          if (el === this.favouritedDropdown) {
            newState.favouritesOpen = true
            break
          }
          el = el.parentElement
        }
      }
      return newState
    })
  }

  handleAddFavourite = event => {
    event.preventDefault()
    const selectedFavouriteName = this.state.newFavouriteName
    this.props.handleAddFavourite(selectedFavouriteName)
    this.setState(() => ({
      newFavouriteName: '',
      selectedFavouriteName,
    }))
  }

  /**
   * Renames the favourite in state
   */
  handleFavouriteNameChange = () => {
    const newValue = this.newFavouriteName.value
    this.setState(() => ({
      newFavouriteName: newValue.replace('_', ''),
    }))
  }

  render() {
    const {
      FavouritesIcon, RemoveFavouriteIcon, OpenIcon, CloseIcon,
      loadFavourite, handleDeleteFavourite,
    } = this.props
    const {favouritesOpen, newFavouriteName, selectedFavouriteName} = this.state
    const favourites = this.props.favourites.map(favourite => (
      <FavouritesItem
        key={`favourite-${favourite.name}`}
        {...favourite}
        loadFavourite={loadFavourite}
        handleDelete={handleDeleteFavourite}
        isSelected={favourite.name === selectedFavouriteName}
        RemoveFavouriteIcon={RemoveFavouriteIcon}
      />
    ))

    return (
      <div className={ClassNames('objectlist-dropdown', {
        open: this.state.favouritesOpen,
      })} ref={el => { this.favouritedDropdown = el }}>
        <button
          className={ClassNames('objectlist-button--dropdown objectlist-button objectlist-button--favourites', {
            open: favouritesOpen,
          })}
          type="button"
          ref={el => { this.favouritesButton = el }}>
          {FavouritesIcon} Favourites
          {favouritesOpen ? CloseIcon : OpenIcon}
        </button>
        <div
          className="objectlist-dropdown__menu">
          {favourites}
          {!!favourites.length && <div className="objectlist-dropdown__divider" />}
          <div className="objectlist-dropdown__item objectlist-dropdown__item--input">
            <form onSubmit={this.handleAddFavourite}>
              <input
                type="text"
                ref={el => { this.newFavouriteName = el }}
                className="objectlist-input objectlist-input__favourite"
                placeholder="Save as new"
                value={newFavouriteName}
                onChange={this.handleFavouriteNameChange} />
              <button type="submit" hidden>Add</button>
            </form>
          </div>
        </div>
      </div>
    )
  }
}

export default Favourites
