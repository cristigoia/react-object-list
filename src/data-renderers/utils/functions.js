/**
 * Gets the value of the cell from the row and keys provided
 * @param  {Object} row  Row supplied
 * @param  {Array|string} keys List of keys used to access the value
 * @return {(Object|string|number)}      Value from row, or row itself if no keys
 */
const getValueFromAccessor = (row, keys) => {
  if (!Array.isArray(keys)) keys = keys.split('.')
  let value = row
  const cloneKeys = [...keys]
  while (value && cloneKeys.length) {
    const thisKey = cloneKeys.shift()
    if (thisKey in value) {
      value = value[thisKey]
    }
  }
  return value
}

export {
  getValueFromAccessor,
}
