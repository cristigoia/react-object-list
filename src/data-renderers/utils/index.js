export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
