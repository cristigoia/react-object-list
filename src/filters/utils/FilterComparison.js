import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
