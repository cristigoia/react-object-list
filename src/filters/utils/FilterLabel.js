import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
