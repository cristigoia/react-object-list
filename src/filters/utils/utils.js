const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
