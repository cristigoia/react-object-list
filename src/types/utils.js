function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
