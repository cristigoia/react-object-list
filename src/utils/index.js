export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
