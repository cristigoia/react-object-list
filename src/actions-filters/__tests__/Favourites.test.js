import React from 'react'
import { snapshotTest } from 'utils/tests'
import Favourites from '../Favourites'

jest.mock('../FavouritesItem', () => 'FavouritesItem')

describe('<Favourites />', () => {
  describe('Snapshot', () => {
    const defaultProps = {
      favourites: [
        {
          name: 'Favourite 1',
          filters: {},
          activeSort: '-id',
          optionalFields: {},
        },
        {
          name: 'Favourite 2',
          filters: {},
          activeSort: '-id',
          optionalFields: {name: true},
        },
      ],
      handleDeleteFavourite: jest.fn(),
      handleAddFavourite: jest.fn(),
    }
    afterEach(() => {
      defaultProps.handleAddFavourite.mockClear()
      defaultProps.handleDeleteFavourite.mockClear()
    })

    it('renders correctly', () => {
      snapshotTest(<Favourites {...defaultProps} />)
    })
  })
})
