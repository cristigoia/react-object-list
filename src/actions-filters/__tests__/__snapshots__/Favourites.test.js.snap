// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favourites /> Snapshot renders correctly 1`] = `
<div
  className="apilist-dropdown"
>
  <button
    className="apilist-button--dropdown apilist-button apilist-button--favourites"
    type="button"
  >
    <i
      className="fa fa-heart-o"
    />
     Favourites
  </button>
  <div
    className="apilist-dropdown__menu"
  >
    <FavouritesItem
      activeSort="-id"
      filters={Object {}}
      handleDelete={[MockFunction]}
      isSelected={false}
      loadFavourite={undefined}
      name="Favourite 1"
      optionalFields={Object {}}
    />
    <FavouritesItem
      activeSort="-id"
      filters={Object {}}
      handleDelete={[MockFunction]}
      isSelected={false}
      loadFavourite={undefined}
      name="Favourite 2"
      optionalFields={
        Object {
          "name": true,
        }
      }
    />
    <div
      className="apilist-dropdown__divider"
    />
    <div
      className="apilist-dropdown__item apilist-dropdown__item--input"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          className="apilist-input apilist-input__favourite"
          onChange={[Function]}
          placeholder="Save as new"
          type="text"
          value=""
        />
        <button
          hidden={true}
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</div>
`;
